### PR TITLE
Refactor the preheader subst map computation in SESE canonicalization. 

### PR DESCRIPTION
**Pull request by bgogul**
Refactoring some code that I find useful for subsequent PRs.

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow linux

**Comment by bgogul** [comment]
@swift-ci please test tensorflow macos

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp [hunk 1]
```diff
@@ -316,6 +316,13 @@ class SingleExitLoopTransformer {
 
   void initialize();
 
+  /// Return a map that captures information about what SILValue should be
+  /// used at the pre-header of the loop for every SILValue in the given
+  /// `values` set. If it cannnot find a suitable SILValue for an entry in
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: typo: cannot

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp [hunk 2]
```diff
@@ -423,20 +443,18 @@ SingleExitLoopTransformer::computeEscapingValues() const {
...
-    }
+llvm::DenseMap<SILValue, SILValue>
+SingleExitLoopTransformer::getPreheaderSubstMap(
+    const SmallPtrSetImpl<SILValue> &values) const {
+  llvm::DenseMap<SILValue, SILValue> result;
+  for (const SILValue value : values) {
+    result[value] = getUndef(value->getType());
   }
+  // Do not eliminate undefs unless requested for.
+  if (!TFNoUndefsInSESE)  return result;
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: Seems to be a double-width space before `return`?

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp [hunk 3]
```diff
@@ -403,14 +423,14 @@ void SingleExitLoopTransformer::initialize() {
 
 llvm::DenseMap<SILValue, SILValue>
 SingleExitLoopTransformer::computeEscapingValues() const {
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: can this be a static method?

**Review by mhong — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp (hunk 3 above)
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: Unfortunately no as it relies on some object state.

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp (hunk 2 above)
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: done.

File: lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp (hunk 1 above)
Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp]: done

**Review by bgogul — COMMENTED** [review]
Thanks for the review.

**Pull request merged**